gnovm/pkg/gnolang/type_check.go

```diff
@@ -215,6 +215,70 @@ func assertAssignableTo(xt, dt Type, autoNative bool) {
 	}
 }
 
+// checkValDefineMismatch checks for mismatch between the number of variables and values in a ValueDecl or AssignStmt.
+func checkValDefineMismatch(n Node) {
+	var (
+		valueDecl *ValueDecl
+		assign    *AssignStmt
+		values    []Expr
+		numNames  int
+		numValues int
+	)
+
+	switch x := n.(type) {
+	case *ValueDecl:
+		valueDecl = x
+		numNames = len(valueDecl.NameExprs)
+		numValues = len(valueDecl.Values)
+		values = valueDecl.Values
+	case *AssignStmt:
+		if x.Op != DEFINE {
+			return
+		}
+
+		assign = x
+		numNames = len(assign.Lhs)
+		numValues = len(assign.Rhs)
+		values = assign.Rhs
+	default:
+		panic(fmt.Sprintf("unexpected node type %T", n))
+	}
+
+	if numValues == 0 || numValues == numNames {
+		return
+	}
+
+	// Special case for single value.
+	// If the value is a call expression, type assertion, or index expression,
+	// it can be assigned to multiple variables.
+	if numValues == 1 {
+		switch values[0].(type) {
+		case *CallExpr:
+			return
+		case *TypeAssertExpr:
+			if numNames != 2 {
+				panic(fmt.Sprintf("assignment mismatch: %d variable(s) but %d value(s)", numNames, numValues))
+			}
+			return
+		case *IndexExpr:
+			if numNames != 2 {
+				panic(fmt.Sprintf("assignment mismatch: %d variable(s) but %d value(s)", numNames, numValues))
+			}
```

Comment:
Okay, maybe we should check to see what happens when we have an expression like `a, ok := s[1]`, where `s` is a string or a slice.